{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (1) Load Processed Data in the Paper\n",
    "\n",
    "Currently, we support norman / adamson / dixit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7bce2d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../')\n",
    "\n",
    "from omnicell.models.gears.pertdata import PertData\n",
    "from omnicell.data.catalogue import Catalogue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2737dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e91ac9e",
   "metadata": {},
   "source": [
    "## (2) Create your own Perturb-Seq data\n",
    "Prepare a scanpy adata object with \n",
    "1. `adata.obs` dataframe has `condition` and `cell_type` columns, where `condition` is the perturbation name for each cell. Control cells have condition format of `ctrl`, single perturbation has condition format of `A+ctrl` or `ctrl+A`, combination perturbation has condition format of `A+B`.\n",
    "2. `adata.var` dataframe has `gene_name` column, where each gene name is the gene symbol.\n",
    "3. `adata.X` stores the post-perturbed gene expression. \n",
    "\n",
    "Here an example using dixit 2016 dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ea020156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 310385 × 8563\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'cell_type'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'gene_id'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "dd = Catalogue.get_dataset_details('repogle_k562_essential_raw')\n",
    "adata = sc.read(dd.path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "425f02ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"condition\"] = adata.obs[\"gene\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9678f7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We relabel NT as ctrl and all other entries as some_entry+ctrl\n",
    "\n",
    "perts = [p for p in adata.obs[\"condition\"].unique() if p != dd.control]\n",
    "adata.obs[\"condition\"] = adata.obs[\"condition\"].replace({dd.control:\"ctrl\"})\n",
    "adata.obs[\"condition\"] = adata.obs[\"condition\"].replace({p:p+\"+ctrl\" for p in perts})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e5d0df6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gem_group</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene_transcript</th>\n",
       "      <th>sgID_AB</th>\n",
       "      <th>mitopercent</th>\n",
       "      <th>UMI_count</th>\n",
       "      <th>z_gemgroup_UMI</th>\n",
       "      <th>core_scale_factor</th>\n",
       "      <th>core_adjusted_UMI_count</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAATCCA-27</th>\n",
       "      <td>27</td>\n",
       "      <td>NAF1</td>\n",
       "      <td>ENSG00000145414</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5449_NAF1_P1P2_ENSG00000145414</td>\n",
       "      <td>NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>11438.0</td>\n",
       "      <td>0.013047</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>14064.512695</td>\n",
       "      <td>k562</td>\n",
       "      <td>NAF1+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-31</th>\n",
       "      <td>31</td>\n",
       "      <td>BUB1</td>\n",
       "      <td>ENSG00000169679</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>935_BUB1_P1P2_ENSG00000169679</td>\n",
       "      <td>BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2</td>\n",
       "      <td>0.179895</td>\n",
       "      <td>5342.0</td>\n",
       "      <td>-1.522247</td>\n",
       "      <td>0.844107</td>\n",
       "      <td>6328.584473</td>\n",
       "      <td>k562</td>\n",
       "      <td>BUB1+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>34</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>9534_UBL5_P1P2_ENSG00000198258</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>0.105287</td>\n",
       "      <td>17305.0</td>\n",
       "      <td>0.384157</td>\n",
       "      <td>1.091537</td>\n",
       "      <td>15853.792969</td>\n",
       "      <td>k562</td>\n",
       "      <td>UBL5+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATAGTC-43</th>\n",
       "      <td>43</td>\n",
       "      <td>C9orf16</td>\n",
       "      <td>ENSG00000171159</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1131_C9orf16_P1P2_ENSG00000171159</td>\n",
       "      <td>C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...</td>\n",
       "      <td>0.099359</td>\n",
       "      <td>30244.0</td>\n",
       "      <td>3.721912</td>\n",
       "      <td>0.948277</td>\n",
       "      <td>31893.619141</td>\n",
       "      <td>k562</td>\n",
       "      <td>C9orf16+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGACAGCGT-28</th>\n",
       "      <td>28</td>\n",
       "      <td>TIMM9</td>\n",
       "      <td>ENSG00000100575</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>8927_TIMM9_P1P2_ENSG00000100575</td>\n",
       "      <td>TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2</td>\n",
       "      <td>0.137623</td>\n",
       "      <td>8407.0</td>\n",
       "      <td>-0.975371</td>\n",
       "      <td>0.868942</td>\n",
       "      <td>9674.979492</td>\n",
       "      <td>k562</td>\n",
       "      <td>TIMM9+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCGTC-45</th>\n",
       "      <td>45</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>ENSG00000100554</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>682_ATP6V1D_P1P2_ENSG00000100554</td>\n",
       "      <td>ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....</td>\n",
       "      <td>0.100272</td>\n",
       "      <td>18350.0</td>\n",
       "      <td>0.428227</td>\n",
       "      <td>1.115052</td>\n",
       "      <td>16456.638672</td>\n",
       "      <td>k562</td>\n",
       "      <td>ATP6V1D+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTCTCG-27</th>\n",
       "      <td>27</td>\n",
       "      <td>CNOT3</td>\n",
       "      <td>ENSG00000088038</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>1718_CNOT3_P1P2_ENSG00000088038</td>\n",
       "      <td>CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2</td>\n",
       "      <td>0.093876</td>\n",
       "      <td>8671.0</td>\n",
       "      <td>-0.633593</td>\n",
       "      <td>0.813253</td>\n",
       "      <td>10662.125000</td>\n",
       "      <td>k562</td>\n",
       "      <td>CNOT3+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTGTGCGG-44</th>\n",
       "      <td>44</td>\n",
       "      <td>METTL3</td>\n",
       "      <td>ENSG00000165819</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>5004_METTL3_P1P2_ENSG00000165819</td>\n",
       "      <td>METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...</td>\n",
       "      <td>0.107983</td>\n",
       "      <td>20568.0</td>\n",
       "      <td>1.054624</td>\n",
       "      <td>0.973352</td>\n",
       "      <td>21131.095703</td>\n",
       "      <td>k562</td>\n",
       "      <td>METTL3+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTGCAGA-14</th>\n",
       "      <td>14</td>\n",
       "      <td>RPL5</td>\n",
       "      <td>ENSG00000122406</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7475_RPL5_P1P2_ENSG00000122406</td>\n",
       "      <td>RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2</td>\n",
       "      <td>0.128225</td>\n",
       "      <td>23568.0</td>\n",
       "      <td>1.676254</td>\n",
       "      <td>1.050055</td>\n",
       "      <td>22444.542969</td>\n",
       "      <td>k562</td>\n",
       "      <td>RPL5+ctrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTACAC-25</th>\n",
       "      <td>25</td>\n",
       "      <td>SEC61B</td>\n",
       "      <td>ENSG00000106803</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>7752_SEC61B_P1P2_ENSG00000106803</td>\n",
       "      <td>SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....</td>\n",
       "      <td>0.095009</td>\n",
       "      <td>13504.0</td>\n",
       "      <td>0.341625</td>\n",
       "      <td>0.855594</td>\n",
       "      <td>15783.187500</td>\n",
       "      <td>k562</td>\n",
       "      <td>SEC61B+ctrl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>310385 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gem_group     gene          gene_id transcript  \\\n",
       "cell_barcode                                                          \n",
       "AAACCCAAGAAATCCA-27         27     NAF1  ENSG00000145414       P1P2   \n",
       "AAACCCAAGAACTTCC-31         31     BUB1  ENSG00000169679       P1P2   \n",
       "AAACCCAAGAAGCCAC-34         34     UBL5  ENSG00000198258       P1P2   \n",
       "AAACCCAAGAATAGTC-43         43  C9orf16  ENSG00000171159       P1P2   \n",
       "AAACCCAAGACAGCGT-28         28    TIMM9  ENSG00000100575       P1P2   \n",
       "...                        ...      ...              ...        ...   \n",
       "TTTGTTGTCTGTCGTC-45         45  ATP6V1D  ENSG00000100554       P1P2   \n",
       "TTTGTTGTCTGTCTCG-27         27    CNOT3  ENSG00000088038       P1P2   \n",
       "TTTGTTGTCTGTGCGG-44         44   METTL3  ENSG00000165819       P1P2   \n",
       "TTTGTTGTCTTGCAGA-14         14     RPL5  ENSG00000122406       P1P2   \n",
       "TTTGTTGTCTTTACAC-25         25   SEC61B  ENSG00000106803       P1P2   \n",
       "\n",
       "                                       gene_transcript  \\\n",
       "cell_barcode                                             \n",
       "AAACCCAAGAAATCCA-27     5449_NAF1_P1P2_ENSG00000145414   \n",
       "AAACCCAAGAACTTCC-31      935_BUB1_P1P2_ENSG00000169679   \n",
       "AAACCCAAGAAGCCAC-34     9534_UBL5_P1P2_ENSG00000198258   \n",
       "AAACCCAAGAATAGTC-43  1131_C9orf16_P1P2_ENSG00000171159   \n",
       "AAACCCAAGACAGCGT-28    8927_TIMM9_P1P2_ENSG00000100575   \n",
       "...                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45   682_ATP6V1D_P1P2_ENSG00000100554   \n",
       "TTTGTTGTCTGTCTCG-27    1718_CNOT3_P1P2_ENSG00000088038   \n",
       "TTTGTTGTCTGTGCGG-44   5004_METTL3_P1P2_ENSG00000165819   \n",
       "TTTGTTGTCTTGCAGA-14     7475_RPL5_P1P2_ENSG00000122406   \n",
       "TTTGTTGTCTTTACAC-25   7752_SEC61B_P1P2_ENSG00000106803   \n",
       "\n",
       "                                                               sgID_AB  \\\n",
       "cell_barcode                                                             \n",
       "AAACCCAAGAAATCCA-27  NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2   \n",
       "AAACCCAAGAACTTCC-31  BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2   \n",
       "AAACCCAAGAAGCCAC-34      UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAACCCAAGAATAGTC-43  C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...   \n",
       "AAACCCAAGACAGCGT-28  TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2   \n",
       "...                                                                ...   \n",
       "TTTGTTGTCTGTCGTC-45  ATP6V1D_+_67826485.23-P1P2|ATP6V1D_+_67826497....   \n",
       "TTTGTTGTCTGTCTCG-27  CNOT3_+_54641532.23-P1P2|CNOT3_-_54641691.23-P1P2   \n",
       "TTTGTTGTCTGTGCGG-44  METTL3_+_21979431.23-P1P2|METTL3_-_21979084.23...   \n",
       "TTTGTTGTCTTGCAGA-14    RPL5_+_93297664.23-P1P2|RPL5_-_93297968.23-P1P2   \n",
       "TTTGTTGTCTTTACAC-25  SEC61B_+_101984577.23-P1P2|SEC61B_-_101984591....   \n",
       "\n",
       "                     mitopercent  UMI_count  z_gemgroup_UMI  \\\n",
       "cell_barcode                                                  \n",
       "AAACCCAAGAAATCCA-27     0.112083    11438.0        0.013047   \n",
       "AAACCCAAGAACTTCC-31     0.179895     5342.0       -1.522247   \n",
       "AAACCCAAGAAGCCAC-34     0.105287    17305.0        0.384157   \n",
       "AAACCCAAGAATAGTC-43     0.099359    30244.0        3.721912   \n",
       "AAACCCAAGACAGCGT-28     0.137623     8407.0       -0.975371   \n",
       "...                          ...        ...             ...   \n",
       "TTTGTTGTCTGTCGTC-45     0.100272    18350.0        0.428227   \n",
       "TTTGTTGTCTGTCTCG-27     0.093876     8671.0       -0.633593   \n",
       "TTTGTTGTCTGTGCGG-44     0.107983    20568.0        1.054624   \n",
       "TTTGTTGTCTTGCAGA-14     0.128225    23568.0        1.676254   \n",
       "TTTGTTGTCTTTACAC-25     0.095009    13504.0        0.341625   \n",
       "\n",
       "                     core_scale_factor  core_adjusted_UMI_count cell_type  \\\n",
       "cell_barcode                                                                \n",
       "AAACCCAAGAAATCCA-27           0.813253             14064.512695      k562   \n",
       "AAACCCAAGAACTTCC-31           0.844107              6328.584473      k562   \n",
       "AAACCCAAGAAGCCAC-34           1.091537             15853.792969      k562   \n",
       "AAACCCAAGAATAGTC-43           0.948277             31893.619141      k562   \n",
       "AAACCCAAGACAGCGT-28           0.868942              9674.979492      k562   \n",
       "...                                ...                      ...       ...   \n",
       "TTTGTTGTCTGTCGTC-45           1.115052             16456.638672      k562   \n",
       "TTTGTTGTCTGTCTCG-27           0.813253             10662.125000      k562   \n",
       "TTTGTTGTCTGTGCGG-44           0.973352             21131.095703      k562   \n",
       "TTTGTTGTCTTGCAGA-14           1.050055             22444.542969      k562   \n",
       "TTTGTTGTCTTTACAC-25           0.855594             15783.187500      k562   \n",
       "\n",
       "                        condition  \n",
       "cell_barcode                       \n",
       "AAACCCAAGAAATCCA-27     NAF1+ctrl  \n",
       "AAACCCAAGAACTTCC-31     BUB1+ctrl  \n",
       "AAACCCAAGAAGCCAC-34     UBL5+ctrl  \n",
       "AAACCCAAGAATAGTC-43  C9orf16+ctrl  \n",
       "AAACCCAAGACAGCGT-28    TIMM9+ctrl  \n",
       "...                           ...  \n",
       "TTTGTTGTCTGTCGTC-45  ATP6V1D+ctrl  \n",
       "TTTGTTGTCTGTCTCG-27    CNOT3+ctrl  \n",
       "TTTGTTGTCTGTGCGG-44   METTL3+ctrl  \n",
       "TTTGTTGTCTTGCAGA-14     RPL5+ctrl  \n",
       "TTTGTTGTCTTTACAC-25   SEC61B+ctrl  \n",
       "\n",
       "[310385 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c66c97a6",
   "metadata": {},
   "source": [
    "### Suggested normalization\n",
    "\n",
    "For raw count data we recommend the following normalization and subsetting to the top 5000 most variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e2168ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata,n_top_genes=5000, subset=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b54492",
   "metadata": {},
   "source": [
    "### Create dataloader\n",
    "\n",
    "GEARS will take it from here. The new data processing takes around 15 minutes for 5K genes and 100K cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fd5e2ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "from omnicell.models.gears.pertdata import PertData\n",
    "\n",
    "pert_data = PertData('./data') # specific saved folder\n",
    "pert_data.new_data_process(dataset_name = 'satija', adata = adata) # specific dataset name and adata object\n",
    "print(f\"Data processed and saved in {pert_data.data_path}\")\n",
    "pert_data.load(data_path = './data/satija') # load the processed data, the path is saved folder + dataset_name\n",
    "print(f\"Data loaded from {pert_data.data_path}\")\n",
    "pert_data.prepare_split(split = 'simulation', seed = 1) # get data split with seed\n",
    "print(f\"Data split with seed 1\")\n",
    "pert_data.get_dataloader(batch_size = 32, test_batch_size = 128) # prepare data loader\n",
    "print(f\"Data loader prepared\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "omnicell",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
